src/iterativesolvers.jl

```diff
@@ -65,16 +65,14 @@ function davidson(A,
   maxsize = size(A)[1]
   actual_maxiter = min(maxiter,maxsize-1)
 
-  if dim(inds(phi)) != maxsize
-    error("linear size of A and dimension of phi should match in davidson")
-  end
+  dim(inds(phi)) != maxsize && throw(DimensionMismatch("linear size of A ($maxsize) and dimension of phi $(dim(inds(phi))) should match in davidson"))
 
   V = ITensor[copy(phi)]
-  AV = ITensor[A(phi)]
+  AV = ITensor[noprime(A*phi)]
```

Comment:
Hi Katie,

Thanks for taking a crack at fitApply. I think the previous version of this was a bit more general, since the operator `A` might not map indices `(i,j,k,...) -> (i',j',k',...)`, which this new version seems to implicitly assume. So I think the logic of mapping the indices back to the proper ones should be handled in a custom function (in the previous version it was the `(pm::ProjMPO)(v::ITensor)`, and I think it wasn't `*` on purpose since `*` would seem to imply that the resulting indices are the `uniqueinds(A,phi)`, instead of the indices of `phi`). 

-Matt

Reply:
Yes, it's not a huge issue but I think Matt's points are all ones to be careful about for the long-term generality of the code. So I'd agree with going back to `A(phi)`.

(Though as a side question, I can't remember whether I defined `A(phi)` or not. Is that overload due to standard usage or compatibility with other Julia libraries? Otherwise just naming that function something like `product` could be even clearer.)

Reply:
The reason I changed it was that there's currently no way to `Davidson` an `ITensor` directly, because it doesn't implement `product` (yet...)

Reply:
Miles: it looks like someone implemented `A(phi)` manually, that is not automatically defined (https://github.com/ITensor/ITensors.jl/blob/master/src/mps/projmpo.jl#L51).

I see... Yeah in the C++ version it required creating a small wrapper around an ITensor that implements `product` and `size` (https://github.com/ITensor/ITensor/blob/v3/unittest/iterativesolvers_test.cc#L11).

I am not sure if it is better to require people to make a small wrapper class like that or overload functions directly on `ITensor`. One concern I have about allowing people to directly pass an ITensor is that something like `size(ITensor)` seems fairly ambiguous, though maybe we can just come up with a different name for it.

I would say that maybe the best approach is for us to provide a standard `ITensorMap` wrapper for people (where it takes an ITensor with indices `(i,j,k,...,i',j',k',...)` and treats it like a matrix). Then you can just call `davidson(ITensorMap(A),...)` for an ITensor `A` than those functions are already available. We can make sure that object works with all of our iterative solver implementations (and perhaps other functions). It could even let us pass some extra type information, like `ITensorMap(A,Hermitian)`, so that we can have a general `eigs` function like Julia has that dispatches on `arnoldi` or `davidson` depending on the symmetry (just something to consider).